src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -767,28 +843,54 @@ private async Task<Dictionary<string, ConfigurationSetting>> LoadSelectedKeyValu
                 };
 
+                List<MatchConditions> matchConditions = new List<MatchConditions>();
+
                 await CallWithRequestTracing(async () =>
                 {
-                    await foreach (ConfigurationSetting setting in client.GetConfigurationSettingsAsync(selector, cancellationToken).ConfigureAwait(false))
+                    AsyncPageable<ConfigurationSetting> pageableSettings = client.GetConfigurationSettingsAsync(selector, cancellationToken);
+
+                    await foreach (Page<ConfigurationSetting> page in _options.PageableManager.GetPages(pageableSettings).ConfigureAwait(false))
                     {
-                        serverData[setting.Key] = setting;
+                        foreach (ConfigurationSetting setting in page.Values)
+                        {
+                            existingData[setting.Key] = setting;
+                        }
+
+                        matchConditions.Add(new MatchConditions { IfNoneMatch = page.GetRawResponse().Headers.ETag });
                     }
                 }).ConfigureAwait(false);
+
+                watchedCollections[new KeyValueIdentifier(KeyFilter.Any, LabelFilter.Null)] = matchConditions;
             }
 
-            foreach (KeyValueSelector loadOption in _options.KeyValueSelectors)
+            foreach (KeyValueSelector loadOption in selectors)
             {
-                IAsyncEnumerable<ConfigurationSetting> settingsEnumerable;
-
                 if (string.IsNullOrEmpty(loadOption.SnapshotName))
                 {
-                    settingsEnumerable = client.GetConfigurationSettingsAsync(
-                        new SettingSelector
+                    var selector = new SettingSelector()
+                    {
+                        KeyFilter = loadOption.KeyFilter,
+                        LabelFilter = loadOption.LabelFilter ?? LabelFilter.Null
+                    };
+
+                    List<MatchConditions> matchConditions = new List<MatchConditions>();
+
+                    await CallWithRequestTracing(async () =>
+                    {
+                        AsyncPageable<ConfigurationSetting> pageableSettings = client.GetConfigurationSettingsAsync(selector, cancellationToken);
+
+                        await foreach (Page<ConfigurationSetting> page in _options.PageableManager.GetPages(pageableSettings).ConfigureAwait(false))
                         {
-                            KeyFilter = loadOption.KeyFilter,
-                            LabelFilter = loadOption.LabelFilter
-                        },
-                        cancellationToken);
+                            foreach (ConfigurationSetting setting in page.Values)
+                            {
+                                existingData[setting.Key] = setting;
+                            }
+
+                            matchConditions.Add(new MatchConditions { IfNoneMatch = page.GetRawResponse().Headers.ETag });
```

Comment:
Can Response or Headers be null?

Reply:
I think if they are null, then the refresh logic kind of breaks. What would we do in that scenario, just throw?

Reply:
Also I guess this kind of applies to anything being null (GetPages, page.Values, this), because in examples on the mslearn docs or github there's no mention of null checks. I think if they're null then something is fully unexpected so I think we just throw

Reply:
Throwing an exception is ok, but null reference exceptions should always be avoided. In this case, we can just make sure that the final etag value is not null. If it is null, we can throw RequestFailedException if SDK isn't already handling this case.

```
ETag serverEtag = page?.GetRawResponse()?.Headers?.ETag ;

if (serverEtag == null)
{
    throw
}

matchConditions.Add(new MatchConditions { IfNoneMatch = serverEtag });
```

Reply:
That makes sense